File: packages/ergoscript/src/eval/negation.ts
/**
 * Negation arm — unary numeric negate (`-x`) on numeric SValues.
 *
 * Result kind equals input kind. `Negate(MIN_K)`: Byte/Short/Int/Long WRAP
 * two's-complement (`-MIN_K === MIN_K`, no error) to match JVM's numeric
 * `negate` (sigma-state `ast/trees.scala:889`); BigInt is 256-bit *checked*
 * and raises `'arith-overflow'` on `-(2^255)` (JVM's 256-bit BigInt overflows
 * too). JVM is canonical here.
 *
 * Sigma-rust ref: ergotree-interpreter/src/eval/negation.rs:16,25
 *   ctx.add_jit_cost(30)?;                       // Negation = Fixed(30)
 *   let input_v = self.input.eval(env, ctx)?;    // eval child after cost
 *   match input_v { Byte/Short/Int/Long/BigInt => checked_neg, ... }
 *
 * ⚠ Deliberate divergence from sigma-rust: its `checked_neg` ERRORS on all five
 * MINs — wrong vs JVM for the 4 machine widths. We wrap them (`_numeric.ts`
 * `maskToKind`) and keep BigInt checked. Routed to the sigma-rust session in
 * `~/projects/santa/prompts/ergots-v5-divergences.md` §A1.
 *
 * Cost-charging order: envelope BEFORE eval-child (sigma-rust line 16 →
 * 17; same posture as LogicalNot / BitInversion).
 *
 * Non-numeric input: sigma-rust returns `EvalError::UnexpectedValue`
 * (negation.rs:35-38). We surface this as `'bin-op-not-numeric'` to
 * match the precedent set by 2c's `LogicalNot` reusing
 * `'bin-op-not-boolean'`. Note: `Negation::try_build`
 * (`ergotree-ir/src/mir/negation.rs:38-50`) rejects non-numeric at
 * build time, so this defensive guard only fires for hand-built MIR
 * nodes (e.g. the inline non-numeric test).
 */

import type { Negation, SValue } from '../mir/types'
import type { Env } from './env'
import type { EvalContext } from './eval-context'
import { EvalError } from './eval-context'
import { evalExpr } from './eval'
import { bigIntToValue, checkRange, isNumeric, maskToKind, valueToBigInt } from './bin-op/_numeric'

const NEGATION_COST = 30

export function evalNegation(e: Negation, env: Env, ctx: EvalContext): SValue {
  ctx.addCost(NEGATION_COST)
  const input = evalExpr(e.input, env, ctx)
  if (!isNumeric(input.kind)) {
    throw new EvalError(
      `Negation: operand kind must be numeric, got '${input.kind}'`,
      'bin-op-not-numeric'
    )
  }
  const negated = -valueToBigInt(input)
  // JVM `Negation.eval` (sigma-state `ast/trees.scala:889`) negates via the
  // numeric `negate`: two's-complement WRAP for the fixed-width types
  // (`-MIN_K === MIN_K`, no error), while BigInt is 256-bit *checked* and throws
  // on `-(2^255)`. sigma-rust `negation.rs:25` uses `checked_neg` for every
  // width and so errors on all five MINs — a divergence from JVM on the 4
  // machine widths. JVM is canonical here (SANTA v5 `Numeric_Negation`; routed
  // to sigma-rust in `~/projects/santa/prompts/ergots-v5-divergences.md` §A1).
  if (input.kind === 'BigInt') {
    checkRange(negated, input.kind, 'arith-overflow')
    return bigIntToValue(input.kind, negated)
  }
  return bigIntToValue(input.kind, maskToKind(negated, input.kind))
}

File: fixture-gen/src/cmds/ergoscript/eval/negation.rs
//! Negation arm — fixtures for `Expr::Negation(...)` evaluation.
//!
//! Sigma-rust ref: `ergotree-interpreter/src/eval/negation.rs:16`
//!   ctx.add_jit_cost(30)?;                      // Negation = Fixed(30)
//!   let input_v = self.input.eval(env, ctx)?;   // eval child AFTER charging cost
//!   match input_v {
//!       Value::Byte(v)   => neg(&v),
//!       Value::Short(v)  => neg(&v),
//!       Value::Int(v)    => neg(&v),
//!       Value::Long(v)   => neg(&v),
//!       Value::BigInt(v) => neg(&v),
//!       _ => Err(EvalError::UnexpectedValue(...)),
//!   }
//!
//! `neg` invokes `checked_neg` per primitive — returns `None` for
//! `MIN_K`, surfaced as `EvalError::ArithmeticException`. On the TS
//! side we reuse `'arith-overflow'` (same code 2c's BinOp.Arith
//! overflow path uses).
//!
//! Cost ordering: envelope charged BEFORE eval-child (sigma-rust line
//! 16 → 17). Same posture as LogicalNot / BitInversion.
//!
//! Coverage:
//!   - 5 kinds × 2 happy values (0 + MAX_K) = 10 success entries.
//!   - 5 overflow entries (`Negate(MIN_K)` per kind) → `'arith-overflow'`.
//!   - 1 cost-limit entry (`jitCostLimit` < per-arm cost) → `'cost-limit-exceeded'`.
//!
//! Non-numeric error case is NOT generated here. `Negation::try_build`
//! (`ergotree-ir/src/mir/negation.rs:38-50`) rejects non-numeric input
//! at build time with `InvalidArgumentError`, so we cannot serialize a
//! malformed tree through the standard path. The TS-side
//! `'bin-op-not-numeric'` assertion is covered by an inline test that
//! calls `evalExpr` directly with a hand-built MIR node (BitInversion
//! precedent — `packages/ergoscript/test/eval/bit-inversion.test.ts`).

use ergotree_interpreter::eval::test_util::try_eval_out;
use ergotree_ir::bigint256::BigInt256;
use ergotree_ir::chain::context::Context;
use ergotree_ir::ergo_tree::{ErgoTree, ErgoTreeHeader};
use ergotree_ir::mir::expr::Expr;
use ergotree_ir::mir::negation::Negation;
use ergotree_ir::mir::unary_op::OneArgOpTryBuild;
use ergotree_ir::mir::value::Value;
use ergotree_ir::serialization::SigmaSerializable;
use num_traits::Bounded;
use serde::Serialize;
use serde_json::{json, Value as JsonValue};
use sigma_test_util::force_any_val;

use super::common::value_to_json;

#[derive(Serialize)]
pub struct NegationFixture {
    pub name: String,
    pub tree_bytes_hex: String,
    pub opts_json: JsonValue,
    /// null for error entries
    pub expected_value_json: JsonValue,
    /// 0 for error entries
    pub expected_cost: u64,
    /// null for success entries
    pub expected_error_code: JsonValue,
}

#[derive(Serialize)]
pub struct NegationFixtureFile {
    pub corpus: &'static str,
    pub entries: Vec<NegationFixture>,
}

fn build_tree(input: Expr) -> anyhow::Result<(ErgoTree, String)> {
    let expr: Expr = Negation::try_build(input)?.into();
    let tree = ErgoTree::new(ErgoTreeHeader::v0(false), &expr)?;
    let hex = hex::encode(tree.sigma_serialize_bytes()?);
    Ok((tree, hex))
}

fn success_entry(name: &str, input: Expr) -> anyhow::Result<NegationFixture> {
    let (tree, hex) = build_tree(input)?;
    let ctx = force_any_val::<Context>();
    let val: Value<'static> = try_eval_out(&tree.proposition()?, &ctx)?;
    Ok(NegationFixture {
        name: name.into(),
        tree_bytes_hex: hex,
        opts_json: json!({}),
        expected_value_json: value_to_json(&val),
        expected_cost: ctx.jit_cost_value(),
        expected_error_code: json!(null),
    })
}

/// Overflow entry — input is numeric (so `try_build` accepts it) but
/// `Negate(MIN_K)` exceeds the signed range. Sigma-rust raises
/// `EvalError::ArithmeticException`; we capture the TS-side code
/// `'arith-overflow'` statically (matching `bin_op_arith.rs` precedent
/// — the `error_entry` helper there doesn't run sigma-rust eval, just
/// builds the tree bytes and assigns the TS error code).
fn overflow_entry(name: &str, input: Expr) -> anyhow::Result<NegationFixture> {
    let (_tree, hex) = build_tree(input)?;
    Ok(NegationFixture {
        name: name.into(),
        tree_bytes_hex: hex,
        opts_json: json!({}),
        expected_value_json: json!(null),
        expected_cost: 0,
        expected_error_code: json!("arith-overflow"),
    })
}

/// Cost-limit entry — `jitCostLimit` set below the per-arm cost (30) so
/// `addCost` overshoots on the very first cost call. Sigma-rust path
/// raises `CostLimitExceeded`; TS-side throws `'cost-limit-exceeded'`.
/// The tree bytes are valid; only the limit makes evaluation fail.
fn cost_limit_entry(name: &str, input: Expr, limit: u64) -> anyhow::Result<NegationFixture> {
    let (_tree, hex) = build_tree(input)?;
    Ok(NegationFixture {
        name: name.into(),
        tree_bytes_hex: hex,
        opts_json: json!({ "jitCostLimit": limit }),
        expected_value_json: json!(null),
        expected_cost: 0,
        expected_error_code: json!("cost-limit-exceeded"),
    })
}

pub fn generate() -> anyhow::Result<NegationFixtureFile> {
    let mut entries = Vec::new();

    // =========================================================================
    // Happy entries — 5 kinds × 2 boundary values (0, MAX_K)
    //   Negate(0) = 0; Negate(MAX_K) = MIN_K + 1 (no overflow).
    // =========================================================================
    entries.push(success_entry("negation_byte_zero", Expr::Const(0i8.into()))?);
    entries.push(success_entry("negation_byte_max", Expr::Const(i8::MAX.into()))?);

    entries.push(success_entry("negation_short_zero", Expr::Const(0i16.into()))?);
    entries.push(success_entry("negation_short_max", Expr::Const(i16::MAX.into()))?);

    entries.push(success_entry("negation_int_zero", Expr::Const(0i32.into()))?);
    entries.push(success_entry("negation_int_max", Expr::Const(i32::MAX.into()))?);

    entries.push(success_entry("negation_long_zero", Expr::Const(0i64.into()))?);
    entries.push(success_entry("negation_long_max", Expr::Const(i64::MAX.into()))?);

    entries.push(success_entry(
        "negation_bigint_zero",
        Expr::Const(BigInt256::from(0i64).into()),
    )?);
    entries.push(success_entry(
        "negation_bigint_max",
        Expr::Const(BigInt256::max_value().into()),
    )?);

    // =========================================================================
    // Overflow entries — Negate(MIN_K) for each kind. checked_neg returns
    // None because |MIN_K| = MAX_K + 1 lies outside the signed range.
    // =========================================================================
    entries.push(overflow_entry(
        "negation_byte_min_overflow",
        Expr::Const(i8::MIN.into()),
    )?);
    entries.push(overflow_entry(
        "negation_short_min_overflow",
        Expr::Const(i16::MIN.into()),
    )?);
    entries.push(overflow_entry(
        "negation_int_min_overflow",
        Expr::Const(i32::MIN.into()),
    )?);
    entries.push(overflow_entry(
        "negation_long_min_overflow",
        Expr::Const(i64::MIN.into()),
    )?);
    entries.push(overflow_entry(
        "negation_bigint_min_overflow",
        Expr::Const(BigInt256::min_value().into()),
    )?);

    // =========================================================================
    // Cost-limit entry — jitCostLimit = 10 < Negation's 30 cost.
    // The very first ctx.addCost(30) overshoots and throws.
    // =========================================================================
    entries.push(cost_limit_entry(
        "negation_cost_limit_exceeded",
        Expr::Const(0i32.into()),
        10,
    )?);

    Ok(NegationFixtureFile {
        corpus: "eval_negation",
        entries,
    })
}
